src/Features/Core/Portable/PullMemberUp/AbstractCodeAndSolutionGenerator.cs

```diff
@@ -0,0 +1,336 @@
+        {
+            if (result.ChangeTargetAbstract)
+            {
+                var abstractTarget = GetAbstractTargetNode(targetNodeSyntax, result.Target, solution);
+                AddMembersToClass(result, abstractTarget, targetNodeSyntax, solution);
+            }
+            else
+            {
+                AddMembersToClass(result, targetNodeSyntax, targetNodeSyntax, solution);
+            }
+
+            RemoveOriginalMembers(result, solution);
+            return solutionEditor.GetChangedSolution();
+        }
+
+        private void AddMembersToClass(
+            AnalysisResult result,
+            SyntaxNode nodeToReplaceTarget,
+            SyntaxNode targetNodeSyntax,
+            Solution solution)
+        {
+            var abstractMembersSymbol = result.MembersAnalysisResults.Where(analysis => analysis.MakeAbstract).
+                Select(analysis => GetAbstractMemberSymbol(analysis.Member));
+
+            var pullUpMembersSymbols = result.MembersAnalysisResults.Where(analysis => !analysis.MakeAbstract).
+                Select(selection => selection.Member).Concat(abstractMembersSymbol);
+
+            var options = new CodeGenerationOptions(generateMembers: false, generateMethodBodies: false, reuseSyntax: true);
+            var membersAddedNode = CodeGenerationService.AddMembers(nodeToReplaceTarget, pullUpMembersSymbols, options: options);
+
+            var editor = EditorMap[solution.GetDocumentId(targetNodeSyntax.SyntaxTree)];
+            editor.ReplaceNode(targetNodeSyntax, membersAddedNode);
+        }
+
+        private void RemoveOriginalMembers(AnalysisResult result, Solution solution)
+        {
+            // If member is marked as make abstract, then don't remove it, just add an abstract declaration to target
+            foreach (var analysisResult in result.MembersAnalysisResults.Where(analysisResult => !analysisResult.MakeAbstract))
+            {
+                foreach (var syntax in SyntaxMap[analysisResult.Member])
+                {
+                    RemoveNode(EditorMap[solution.GetDocumentId(syntax.SyntaxTree)], syntax, analysisResult.Member);
+                }
+            }
+        }
+        
+        private async Task InitializeEditorsAndSyntaxes(
+            AnalysisResult result,
+            SyntaxNode targetSyntaxNode,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            // Members and target may come from different documents, and some of them may also share the same document.
+            // So EditorMap is used to save and group all the editors will be used.
+            EditorMap = new Dictionary<DocumentId, DocumentEditor>();
+            // One member may have multiple syntaxNodes (e.g partial method).
+            // SyntaxMap is used to find the syntaxNodes need to be changed more easily.
+            SyntaxMap = new Dictionary<ISymbol, IEnumerable<SyntaxNode>>();
```

Comment:
you're mutating shared state.  If you want a place to store data that is safe to mutate across an operation, you need to make a new type to put it in.  I can help explain more,but i'm a little tired, so let's do it either over gitter, or maybe someone on the team can help you out with that if you can't reach me.